### ISSUE TITLE
White list of enabled keys by URL

### ISSUE THREAD

**Issue by borislt** [philc/vimium#3190]
The topmost feature configuration in Vimium's settings is: 'Excluded URLs and keys'. 
Which is basically a blacklist.
But to enable only a couple of keys (e.g. scroll, tab switch) - you have to go through and disable all other keys. It's tedious.
Would be nice to have also a White list of enabled keys by URL.

**Comment by davidgenesiscruz** [comment]
+1
I use keyboard shortcuts, for example, on video streaming sites to control playback. Vimium shortcuts would sometimes interfere with them so I disable Vimium on those URLs, but it would be nice if I could just keep the bare essentials like opening tabs.

**Comment by afilp** [comment]
I also need this a lot. 

`Youtube` is one example. The `F` key there is "Full Screen". I had to exlcude "vimium" all together for Youtube, because I really need the "Full Screen" shortcut also.

**Comment by smblott-github** [comment]
> Would be nice to have also a White list of enabled keys by URL.

Yes.  But it would also have to be obvious for new users.  And when you bundle it together with the feature request that you can also exclude (or include) modified keys, it all becomes quite complicated.

That's why I've never taken it on.  I just list the keys I want to use from the page.

@davidgenesiscruz  and @afilp ... You might be looking for insert mode (`i`), or perhaps the `passNextKey` command.

**Comment by afilp** [comment]
@smblott-github Thanks! How about an interface similar to this?

![image](https://user-images.githubusercontent.com/7850073/51046146-fe462680-15cd-11e9-987d-1ff36ea6c266.png)

Notes:
- The `Keys` input box will be **disabled** when `All Keys` value is selected. This will have (internally) the same representation as using `Specific Keys` with no specific value in `Keys` (so same as today, but with a new explicit option too).
- The `All Keys Except` option allows you to add keys that "survive" any exclusion from other rules (i.e. "white list").

You may also use colors in the background of `Keys` (red/green) to show their purpose:

![image](https://user-images.githubusercontent.com/7850073/51046489-dc00d880-15ce-11e9-848f-530bce3dfd5a.png)

And this is the styling with the "All Keys" showing the `Keys` input as disabled.

**Comment by off-by-one** [comment]
I second this - my current workaround is to list all keys I remember and then remove the keys I want to use. This is especially true on websites where I use vim controls internally (cocalc, overleaf), which mean I need very restrictive whitelists to keep vimium from interfering with the site.

**Comment by gdh1995** [comment]
You may be interested in this idea: https://github.com/philc/vimium/issues/3272#issuecomment-475296695 , and I've implemented it in my customized version of Vimum, [Vimium C](https://github.com/gdh1995/vimium-c).

**Comment by ght** [comment]
I like this feature proposal and would use this. I find @afilp's UI mockup unnecessarily complex though:
1. there are really only 2 options, not three: whitelist and blacklist. "All Keys" is already covered by selecting whitelist and leaving it empty
2. not avoiding double negatives does the opposite of reducing confusion; Instead of "Exclude all keys except …", I suggest "Allow only …"

I'm not good with mockups, but I suggest to keep this as simple as possible:
- a simple toggle icon added to the current UI
- two states: blacklist(default)/whitelist
- when toggled:
  - the icon changes
  - placeholder text changes  ("Exclude keys"/"Allow keys"), show that as title text/tooltip on the icon as well
  - maybe the color changes too as @afilp proposed

Maybe "Enable" would be better than "Allow", I'm not a native speaker.

**Comment by actionless** [comment]
possibly duplicate of https://github.com/philc/vimium/issues/2890

**Comment by gaerfield** [comment]
For everybody else finding its way here, this is my (incomplete?) list of all keys I got identified: 
```
jkgGuhlzHzLryypPgugUivVgifFmoObBTeE/nNHLtJK^yxXW?
```
Until solved, copy&paste this and remove keys you like to use.

**Comment by atsalolikhin-spokeo** [comment]
I would like to be able to use GitHub's `y` hotkey to get the permalink when on GitHub (see https://docs.github.com/en/github/managing-files-in-a-repository/getting-permanent-links-to-files)

I don't want to turn off Vimium on GitHub, because I can't live without Vimium.  :)  For example, I use `yy` to copy the current link.  So... is there some way I can get the GitHub permalink through Vimium?  (Not sure if this belongs here or in its own issue.)

**Comment by gdh1995** [comment]
@atsalolikhin-spokeo You may try adding exclusion rules to always pass some keys to GitHub, or binding a key to `passNextKey` to pass a key once in a while.

**Comment by atsalolikhin-spokeo** [comment]
@gdh1995 : thanks for replying!  I considered writing an exclusion rule to always pass `y` to GitHub, but that would break the Vimium `yy` key combo, which I use to copy the URL of the current page.  

I wasn't aware I could bind a key to `passNextKey`, there is no mention of that on the Vimum Options page. 

Aha! I found https://github.com/philc/vimium/wiki/Tips-and-Tricks  Yay!  :)  Thank you.  :)

**Comment by CMCDragonkai** [comment]
I'm starting to use Linear, and it has its own keyboard shortcuts, would be nice if I can specify keys to allow, instead of keys to disallow.

**Comment by maksimmuravev** [comment]
> For everybody else finding its way here, this is my (incomplete?) list of all keys I got identified:
> 
> ```
> jkgGuhlzHzLryypPgugUivVgifFmoObBTeE/nNHLtJK^yxXW?
> ```
> 
> Until solved, copy&paste this and remove keys you like to use.

Vimium still overrides <ESC>.

**Comment by philc** [comment]
Adding a reference to #1560, which I think was the original issue filed for this, and has some discussion.

**Comment by philc** [comment]
Empirically, over the years this is one of the more requested features, so I'll take another pass to see if a reasonable solution is viable.

**Comment by souvlakeeb** [comment]
Thank you very much! I'm sure a lot of people will be happy about this. 

My first thought when trying to work this problem was to define a negative pattern similar to bash's `!(*all-except-this*)`. The problem with that is `!`, `(`, and `)` can be bound. But, still, I felt it important to share for your consideration.
